feat(gui): Add user option to adjust scale of game fonts in relation to resolution

Mauller opened the pull request:
This PR adds user based font scaling adjustment to the game.

The value can be set within the options.ini by setting `ResolutionFontAdjustment = 10` etc, where the value is a percentage increase, the default being 0% of the original scale.

This value acts on the normal ini based resolution scaling to give some user flexibility.

----
TODO:

- [x] Replicate in generals

xezon reviewed (commented):
I suggest to merge this change before the other, so we can experiment with this.

Mauller commented on GeneralsMD/Code/GameEngine/Source/GameClient/GUI/GUICallbacks/Menus/OptionsMenu.cpp:
```diff
@@ -828,6 +828,20 @@ Int OptionPreferences::getGameTimeFontSize(void)
 
+Real OptionPreferences::getUserFontScale(void)
+{
+	OptionPreferences::const_iterator it = find("UserFontScale");
```
The idea behind this one was to be a bit of an accessibility feature for people who might find the font too small etc.

xezon commented on GeneralsMD/Code/GameEngine/Source/GameClient/GUI/GUICallbacks/Menus/OptionsMenu.cpp:
```diff
@@ -828,6 +828,20 @@ Int OptionPreferences::getGameTimeFontSize(void)
+		return 1.0f;
+
+	Real fontScale = (Real)atof(it->second.str()) / 100.0f;
+	if (fontScale < 0.5f)
```
0 would mean that there is no scaling in higher resolutions. I think it is ok.

Mauller replied:
i would have to change the implementation for zero to not change the size, at the moment it's scaling the value as a percentage of the value.

Mauller replied:
tweaked it so zero is a valid value that does not scale the font

Mauller commented:
Rebased, removed the other font scaling stuff and tweaked based on feedback.

This now will not scale the font if the value is set to zero, which i have set as the default.

Mauller commented:
> This is not what I meant :)
> 
> `adjustFactor = 1.0f + (adjustFactor-1.0f) * m_resolutionFontSizeAdjustment * TheGlobalData->m_resolutionFontAdjustment;` was fine, because if `TheGlobalData->m_resolutionFontAdjustment` is 0, then the adjust factor simply will be one: no resolution scaling, text has same size as with 800 x 600
> 
> `* (1 + TheGlobalData->m_resolutionFontAdjustment)` is misleading, because 1 becomes 2 (doubling).

Yeah i thought you meant that zero would result in the extra user option not having any effect on the current scaling.
So the game essentially just scales like it normally would.

Also this is the game with just the raw Font sizes, so m_resolutionFontAdjustment at zero
<img width="567" height="479" alt="image" src="https://github.com/user-attachments/assets/e649378e-c82e-448f-80ad-a35ce4fc5988" />

Mauller commented:
So tweaked this now so by default for this PR the user based setting does not adjust the default from languages.ini

Mauller commented on GeneralsMD/Code/GameEngine/Source/GameClient/GlobalLanguage.cpp:
```diff
@@ -169,10 +169,15 @@ void GlobalLanguage::init( void )
+void GlobalLanguage::reset( void )
+{
+	if (TheGlobalData->m_resolutionFontAdjustment >= 0.0)
+		m_resolutionFontSizeAdjustment = TheGlobalData->m_resolutionFontAdjustment;
```
it's why i used "UserFontAdjustment" as the naming originally.

Mauller replied:
Tweaked it

Mauller commented on GeneralsMD/Code/GameEngine/Source/GameClient/GlobalLanguage.cpp:
```diff
@@ -169,10 +169,15 @@ void GlobalLanguage::init( void )
+void GlobalLanguage::reset( void )
+{
+	if (TheGlobalData->m_resolutionFontAdjustment >= 0.0)
+		m_resolutionFontSizeAdjustment = TheGlobalData->m_resolutionFontAdjustment;
```
yeah i missed that part of it, i considered overwriting the original value since the function for font adjustment gets hit a lot, but one extra function call in this instance shouldn't hurt.

Mauller commented:
Tweaked based on some of the feedback.

Mauller commented:
Tweaked now so the GlobalLanguage reads from the options.ini and loads the value locally.

The options menu then reads from the options ini and sets the value in GlobalData

Mauller commented:
will replicate to generals later if we think this is good to go now.

xezon commented on GeneralsMD/Code/GameEngine/Include/GameClient/GlobalLanguage.h:
```diff
@@ -101,9 +101,11 @@ class GlobalLanguage : public SubsystemInterface
 
 	//UnicodeString	m_unicodeFontNameUStr;
 
+	float getResolutionFontSizeAdjustment();
```
const

xezon commented on Generals/Code/GameEngine/Source/GameClient/GlobalLanguage.cpp:
```diff
@@ -185,10 +192,18 @@ void GlobalLanguage::parseFontFileName( INI *ini, void * instance, void *store,
 	monkey->m_localFonts.push_front(asciiString);
 }
 
+const float GlobalLanguage::getResolutionFontSizeAdjustment( void )
```
const

Mauller replied:
not sure what you mean by this since you already asked for it to be const and i made it const.

xezon replied:
Sorry, should have been more specific :-D

Returning const value only make sense for pointers or references :-)

xezon commented on GeneralsMD/Code/GameEngine/Source/GameClient/GUI/GUICallbacks/Menus/OptionsMenu.cpp:
```diff
@@ -1354,6 +1369,17 @@ static void saveOptions( void )
+	//-------------------------------------------------------------------------------------------------
+	// Set User Font Scaling Percentage
+	val = pref->getResolutionFontAdjustment() * 100.0f; // TheSuperHackers @todo replace with options input when applicable
+	if (val >= 0)
```
if this is -1 default, then this is not saved to Options.ini? Is this intended?

Mauller replied:
The checks there more for later when we have UI elements to set the value instead of it being taken from prefs directly as it currently is.

ElTioRata commented:
GUIEdit doesn't start anymore after this PR.

Mauller commented:
It opens and runs perfectly fine for me when building it locally.

Mauller commented:
This seems to only be a VC6 issue

Mauller commented:
Think i may have figured it out now, global languages is initialised before global data in guiedits init function.
just re-arranged to test.
